Mountain Lion Scrollers

avaidyam opened the pull request:
Pretty massive update to bring the scrollers' look to match Lion's (with a slight border), as well as Mountain Lion's (the scroller track expansion). I've also updated the API a bit.

avaidyam committed:
Convert TUIScrollKnob to a pretty TUIScroller.

avaidyam committed:
TUIScroller now expands on mouse over.

avaidyam committed:
Add knob border colors and change colors for indicator style.

avaidyam committed:
ScrollKnob -> Scroller (Matches AppKit)

avaidyam committed:
TUIScroller animations + 10.8 checks.

jspahrsummers commented on lib/UIKit/TUIScrollView.m:
```diff
@@ -353,139 +356,161 @@ - (void)setResizeKnobSize:(CGSize)s
-	CGFloat knobSize = 12;
+	
+	CGFloat verticalScrollerSize = self.verticalScroller.expanded ? 15 : 11;
+	CGFloat horizontalScrollerSize = self.horizontalScroller.expanded ? 15 : 11;
```
These should use the constants defined in TUIScroller.m, so any changes only have to be made to one location.

jspahrsummers commented on lib/UIKit/TUIScrollView.m:
```diff
@@ -353,139 +356,161 @@ - (void)setResizeKnobSize:(CGSize)s
+	CGFloat horizontalScrollerStart = 0;
+	
+	CGFloat verticalScrollerOffset = isAtleastLion ? 0 : ((hVisible ? verticalScrollerSize : 0) + resizeKnobSize.height);
+	CGFloat horizontalScrollerOffset = isAtleastLion ? 0 : ((vVisible ? horizontalScrollerSize : 0) + resizeKnobSize.width);
```
Please change this so there aren't nested ternary expressions (use one or two `if` statements instead).

avaidyam replied:
Hurrr. A bad habit of mine.

jspahrsummers commented on lib/UIKit/TUIScrollView.m:
```diff
@@ -353,139 +356,161 @@ - (void)setResizeKnobSize:(CGSize)s
+		self.verticalScroller.frame = newVScrollerRect;
+		self.horizontalScroller.frame = newHScrollerRect;
+	};
+	if(!animated) {
```
This check can be moved into the `animations` block, so that the `redraw` invocations don't need to be duplicated.

avaidyam replied:
Nifty! Thanks.

jspahrsummers commented on lib/UIKit/TUIScroller.h:
```diff
@@ -15,29 +15,17 @@
-		unsigned int scrollIndicatorStyle:2;
-		unsigned int flashing:1;
-	} _scrollKnobFlags;
-}
```
:metal: Thanks for removing these!

avaidyam replied:
It's a pleasure!

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,385 @@
+#import "NSColor+TUIExtensions.h"
+
+extern BOOL isAtleastLion;
+extern BOOL isAtleastMountainLion;
```
These should be declared in a private header or something. Referencing them via `extern` here instead of importing a declaration will make it harder to understand where any hypothetical breakage occurs.

avaidyam replied:
How about we move them into a `TUIView` private header? This way, any `TUIView` that needs to know can just import the header and automatically know.

jspahrsummers replied:
It seems like only `TUIScrollView`-related things should care. I'd like to avoid OS conditionalizing as much as possible, so let's start by just exposing it in a `TUIScrollView` private header.

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,385 @@
+	} _scrollerFlags;
+}
+
+@property (nonatomic, assign) BOOL knobHidden;
```
This property should have a getter of `isKnobHidden`.

avaidyam replied:
Missed that one, but I generally do that.

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,385 @@
+
+- (id)initWithFrame:(CGRect)frame {
+	if((self = [super initWithFrame:frame])) {
+		_knob = [[TUIView alloc] initWithFrame:CGRectMake(0, 0, 12, 12)];
```
Shouldn't this use `TUIScrollerDefaultWidth` and `TUIScrollerMinimumKnobSize`?

avaidyam replied:
It's gonna be set either way when `_updateKnob` is called, so there's not much point right now. I'd rather use `CGRectZero`.

jspahrsummers replied:
:+1: Let's do that.

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,385 @@
+		_knob = [[TUIView alloc] initWithFrame:CGRectMake(0, 0, 12, 12)];
+		self.knob.userInteractionEnabled = NO;
+		self.knob.layer.borderWidth = 1.0f;
+		self.scrollIndicatorStyle = TUIScrollViewIndicatorStyleDark;
```
Seems like this should be `TUIScrollViewIndicatorStyleDefault`.

avaidyam replied:
Oops, debugging leftovers.

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,385 @@
+}
+
+- (void)_refreshKnobTimer {
+	if([NSScroller preferredScrollerStyle] == NSScrollerStyleOverlay) {
```
Can you please flip this condition and return early, instead of putting this whole method in a conditional?

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,385 @@
+	return self.bounds.size.height > self.bounds.size.width;
+}
+
+#define TUIScrollerCalculations(OFFSET, LENGTH, MIN_KNOB_SIZE) \
```
This should be a `static` function with fully documented inputs and outputs. This is impossible to read.

avaidyam replied:
I know, but it uses more than one variable from inside that. It'd be hard to change that.

jspahrsummers replied:
In that case, the static function can accept pointer arguments which it will set upon return, similar to `CGRectDivide`, or how `NSErrors` are typically returned in Cocoa.

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,385 @@
+	[super mouseUp:event];
+}
+
+#define TUIScrollerCalculationsReverse(OFFSET, LENGTH) \
```
This should be a `static` function with fully documented inputs and outputs. This is impossible to read.

avaidyam replied:
Same as above-- uses more than one variable "output".

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,385 @@
+		frame.origin.x = 0.0;
+		frame.origin.y = knobOffset;
+		frame.size.height = MIN(2000, knobLength);
+		frame.size.width = self.expanded ? TUIScrollerExpandedWidth : TUIScrollerDefaultWidth;
```
This can be collapsed into a `CGRectMake`.

avaidyam replied:
Sure!

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,385 @@
+		frame.origin.x = knobOffset;
+		frame.origin.y = 0.0;
+		frame.size.width = MIN(2000, knobLength);
+		frame.size.height = self.expanded ? TUIScrollerExpandedWidth : TUIScrollerDefaultWidth;
```
This can be collapsed into a `CGRectMake`.

avaidyam replied:
As above.

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,385 @@
+		frame.origin.y = knobOffset;
+		frame.size.height = MIN(2000, knobLength);
+		frame.size.width = self.expanded ? TUIScrollerExpandedWidth : TUIScrollerDefaultWidth;
+		frame = ABRectRoundOrigin(CGRectInset(frame, 2, 4));
```
These constants should be named as well.

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,385 @@
+		frame.origin.y = 0.0;
+		frame.size.width = MIN(2000, knobLength);
+		frame.size.height = self.expanded ? TUIScrollerExpandedWidth : TUIScrollerDefaultWidth;
+		frame = ABRectRoundOrigin(CGRectInset(frame, 4, 2));
```
These constants should be named as well.

avaidyam replied:
They're padding values.

jspahrsummers replied:
We can still give them a name, especially since they're flipped for horizontal and vertical scrollers.

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,385 @@
+						   [NSColor colorWithCalibratedWhite:0.95 alpha:0.85]];
+	NSArray *lightTrack = @[[NSColor colorWithCalibratedWhite:0.15 alpha:0.85],
+							[NSColor colorWithCalibratedWhite:0.20 alpha:0.85]];
+	BOOL drawLight = self.scrollIndicatorStyle == TUIScrollViewIndicatorStyleLight;
```
Shouldn't the light track also be used for the default style?

avaidyam replied:
That's what it does at the moment. I think I need to rename these variables-- by `darkTrack`, I mean the `track for the dark scroller`.

jspahrsummers replied:
How about `darkScrollerTrack` and `lightScrollerTrack`?

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,385 @@
+	animation.values = @[@0.5f, @0.2f, @0.0f];
+	[self.knob.layer addAnimation:animation forKey:@"opacity"];
+	
+	[self performSelector:@selector(_endFlashing) withObject:nil afterDelay:(duration - 0.01)];
```
This should occur in the animation's delegate callback instead.

avaidyam replied:
Got it!

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,385 @@
+
+- (void)mouseEntered:(NSEvent *)event {
+	_scrollerFlags.hover = 1;
+	[self _updateKnobAlphaWithSpeed:0.08];
```
This should be pulled out into a named constant.

avaidyam replied:
But I'm not 100% sure what these constants mean. I know it's a speed, but _why_ is it 0.08 is my question. I'll still pull them out if I can....

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,385 @@
+
+- (void)mouseExited:(NSEvent *)event {
+	_scrollerFlags.hover = 0;
+	[self _updateKnobAlphaWithSpeed:0.25];
```
This should be pulled out into a named constant.

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,385 @@
+	_mouseDown = [self localPointForEvent:event];
+	_knobStartFrame = self.knob.frame;
+	_scrollerFlags.active = 1;
+	[self _updateKnobAlphaWithSpeed:0.08];
```
This should be pulled out into a named constant.

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,385 @@
+
+- (void)mouseUp:(NSEvent *)event {
+	_scrollerFlags.active = 0;
+	[self _updateKnobAlphaWithSpeed:0.08];
```
This should be pulled out into a named constant.

jspahrsummers commented:
:postbox: I know some of this code was pre-existing, but I've made comments on it as if it was all new, because this stuff should be fixed if we're touching the code anyways.

In addition to the code comments, the ML-like behaviors are pretty wonky, as can be seen in the example app. In particular, the track expansion often causes the scroll knob to jump around. I think this work needs to be tested much more thoroughly and with as many different configurations as possible – this is something that could be done in the example app.

Can you please also merge from master?

jwilling commented:
:+1: Much better than the previous scrollers. However, they do not look exactly like the built-in ones. Something about the outer border doesn't look quite right. Might be worth a look.

choco commented:
I'm testing on a rMBP and as noted by @jwilling the outer border doesn't look right to me either :/ Actually I don't think there's such border in ML scrollers, or maybe only on the lighter version?  Also I I noticed that if the know is expanded and I try to scroll, the know get's shrinked right away (not standard ML behavior), and that the knob disappears to fast when we need to hide it, TUIScrollerStateChangeSpeed needs some love :)

jwilling commented:
@cHoco I'm on a rMBP as well and perhaps that's where the problem is. My scrollers don't always show a border. I have yet to figure out when it decides it needs it.

avaidyam commented:
Could anyone help determine when the knob as a border? That's been a _female dog_ for me to get right. I'm fixing up what @jspahrsummers noted at the moment, so I guess I'll check out that disappearance glitch. I've been having it too, and I'm not ready to :shipit: [ship it squirrel] if it's gonna hate on me and be all glitchy.

jwilling commented:
@galaxas0 When the slider track gets shown, it looks like OS X hides the border. On lighter backgrounds, it seems like it hides the border as well.

avaidyam commented:
Yeah, I've done criteria A. But if I navigate to phandroid.com (Oh yes, Android > iOS, as awesome as Mac is) I can see the light border clearly. But when I navigate to google.com, there's _also_ a border there. :confused:  

## 

Aditya.

On Monday, October 15, 2012 at 8:32 AM, Jonathan Willing wrote:

> @galaxas0 (https://github.com/galaxas0) When the slider track gets shown, it looks like OS X hides the border. On lighter backgrounds, it seems like it hides the border as well.
> 
> —
> Reply to this email directly or view it on GitHub (https://github.com/github/twui/pull/109#issuecomment-9449331).

bestwnh commented:
@galaxas0 I can see the border all the time with my ML(10.8.2) scroller. But I don't see that in your example project. I don't think the OS X will hide the border.

avaidyam commented:
Thanks for your report! I've been doing some testing, and I'll upload a few sample images of the default NSScroller.

avaidyam committed:
Style fixes and refactoring.

avaidyam commented:
@jspahrsummers I've fixed whatever you noted.

Can anyone tell me what border colors I should use and where? They're driving me nuts!

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,417 @@
+	CGRect frame = CGRectZero;
+	if(self.vertical) {
+		frame = CGRectMake(0.0, knobOffset, self.updatedScrollerWidth, knobLength);
+		frame = ABRectRoundOrigin(CGRectInset(frame, 2, 4));
```
These still need to be named. See my previous reply.

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,417 @@
+	BOOL drawLight = self.scrollIndicatorStyle == TUIScrollViewIndicatorStyleLight;
+	NSArray *lightTrack = @[[NSColor colorWithCalibratedWhite:0.90 alpha:0.85],
+							[NSColor colorWithCalibratedWhite:0.95 alpha:0.85]];
+	NSArray *darkTrack = @[[NSColor colorWithCalibratedWhite:0.15 alpha:0.85],
```
These still need better names. See my previous reply.

avaidyam replied:
Fixed.

jspahrsummers commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,417 @@
+	[self _updateKnobAlphaWithSpeed:TUIScrollerRefreshSpeed];
+	
+	// Normal knob dragging scroll.
+	// We can't use hitTest because userInteractionEnabled = NO.
```
Why does user interaction need to be disabled?

avaidyam replied:
I'm not too sure; I'd ask the people who wrote this class.  Perhaps it was to pass events through to the track, so new methods would not have to be implemented. I didn't fiddle around with that though.

avaidyam committed:
Name changes.

avaidyam committed:
Merge remote-tracking branch 'upstream/master' into upstream-ML-scrollbars

Conflicts:
	lib/UIKit/TUIScrollKnob.m

avaidyam committed:
Fix scroller outline. Switched to shadow from border.

avaidyam committed:
Merge remote-tracking branch 'upstream/master' into upstream-ML-scrollbars

avaidyam committed:
Fixed animation glitches.

avaidyam commented:
@jspahrsummers This should be good to go now!
@jwilling Cold you test it out on a Retina display?

jwilling commented:
Couple of issues I noticed.
- Moving the mouse back into the bounds of the scroll view should not trigger a show of the scroll bars.
- The opacity of the scroll knob seems to be a little bit darker than the built-in one.
- If you scroll really slowly, the knob jumps up a pixel on one edge while the other one stays static. In turn, if you scroll up a small bit more the other edge "catches up" a pixel. This is probably due to some sort of rounding error that isn't keeping the height of the knob constant.

avaidyam committed:
Fixed scroller jumping.

avaidyam committed:
Fixed extraneous flashing scrollers.

avaidyam commented:
@jwilling @jspahrsummers Great, fixed those issues as well. 
The jumping issue was with `CGRectIntegral`... I guess Loren had a good reason for using `ABRoundOrigin`.

jspahrsummers commented:
This still needs some polish.

In particular, with ML-style expanding scrollers, the expansion of the scroll track causes the knob to briefly jump off-screen to the right before coming back. If the scroller is also moving vertically, the weirdness is compounded.

avaidyam commented:
Could you provide an example? I'm trying quite hard to reproduce this effect, but it's not working.

jspahrsummers commented:
Nice blurry screen recording of the behavior I was describing: https://dl.dropbox.com/u/92083149/twui-scroller-oddities.mov

avaidyam committed:
Don't allow both expanded tracks to expand at once.

Preview.app is a good example. Try playing with both tracks and one of
them hides, when the other is shown right after.

avaidyam committed:
Merge remote-tracking branch 'upstream/master' into upstream-ML-scrollbars

Conflicts:
	lib/UIKit/TUIScrollView.m

avaidyam commented:
Ah, I misunderstood you! Okay, so how the scroller squashes off-screen then expands back into place. (Not really, but this is how I see it...)
Got it. I'll fix that-- and I've also fixed a very interesting glitch where both scroller tracks would display on screen at once. Preview.app only shows one track at a time. If you mouse over on vertical track, and then quickly to horizontal, the vertical track disappears. So this is now fixed in TwUI as well.

avaidyam committed:
Fix the scroller positioning and animation issues.

Anchor the scroller and set its position without animation first.
Make sure the scroller's rect isn't below/above the track.

avaidyam commented:
@jspahrsummers Fixed!  :mailbox:

jspahrsummers commented:
Now the scroller expansion doesn't animate at all, which is also wrong.

avaidyam commented:
@jspahrsummers I'm not sure how to disable animations for everything _but_ width/origin frame changes. It ends up becoming messy. Could you help me with that?

jwilling commented:
@galaxas0 Either use `- (id<CAAction>)actionForKey:(NSString *)key`, or preferrably set the layer's actions to a dictionary  with the key pairs containing the specified key paths with objects of `NSNull`.

For example, to disable actions for a bounds change, you can do something along the lines of the following in some sort of initialization method:

```
NSDictionary *actions = @{ @"bounds" : [NSNull null] };
self.layer.actions = actions;
```

avaidyam commented:
@jwilling Could you elaborate on the first method? I've never heard of that method being used like that.

jwilling commented:
@galaxas0 http://developer.apple.com/library/mac/#documentation/cocoa/conceptual/CoreAnimation_guide/Articles/Actions.html

avaidyam commented:
@jwilling It seems like a more complicated issue than that: as we expand and set the frame, the layer's `position.x` is being moved left, and the layer's `bounds.size.width` is being moved right, so it's acting wonky. Both methods above didn't work for me.

I've tried to use `anchorPoint` on both `self.layer` and `self.knob.layer` to keep width changes pushed left, but it's not working either...

avaidyam commented:
@jwilling Just kidding, I put the code in the wrong place, and now it works. But there's another issue I've to fix...

avaidyam commented:
Alright, the wonky behavior is gone now, and it expands perfectly well, but now, for some reason, when it first shows, it's off-position. Anyone know why?

avaidyam commented:
@jwilling @jspahrsummers After a bit of debugging, I've fixed all the issues. It should be good to go now. (For real!) :mailbox_with_mail:

avaidyam commented:
@jspahrsummers I've fixed the issue :mailbox_with_mail:

jwilling commented:
One thing to note is that on OS X when the content is pulled past the boundaries, the scroll knob does not travel past the top of the bounds. It shrinks instead.

This probably shouldn't be happening:

![](http://i.imgur.com/axgmu.png)

avaidyam commented:
That was pre-existing behavior-- I've adjusted it to the extent that it doesn't go upwards. This issue no longer exists in this PR. (Or shouldn't.)

jwilling commented:
The screenshot taken above was from the code in your pull request.

avaidyam commented:
Alright, I'll fix that up then.

jwilling commented:
Don't forget about the bottom as well.

![](http://i.imgur.com/xWY8z.png)

avaidyam commented:
@jwilling @jspahrsummers There aren't anymore issues with this, as far as I can see, and they look and act, if not exactly like, close to the original Lion and ML scrollers... Is this okay to be merged?

jwilling commented:
This is not ready to be merged. Open up the scroll view example and notice the flashing borders and visual glitches.

avaidyam commented:
@jwilling I don't notice any flashing borders, and the only glitch I see is the horizontal scroller being stuck... I can fix that too.

avaidyam committed:
Fix improper scroller positioning and scroller seizures.

jwilling commented:
How about you take a look at this screen recording? http://cl.ly/KY1Y

I immediately notice two issues here.
- The horizontal scrollbar has some sort of animation applied to it, hence the appearance of lag.
- When both knob tracks are shown, they flash.

jwilling commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,461 @@
+		else if(knobOffset + knobLength > trackBounds.size.width)
+			knobOffset = trackBounds.size.width - knobLength;
+		else if(knobOffset < trackBounds.origin.x)
+			knobOffset = trackBounds.origin.x;
```
This could be condensed a good deal by using a one-line ternary statement. No need to duplicate so much code.

jspahrsummers replied:
A ternary expression on what? There are several conditions here, although I agree they're duplicative.

jwilling replied:
That's what happens when attempting to look at code late at night. Must've glazed over the previous width/height statement. I also just realized GH's coding conventions don't allow inline ternary statements, so that one went out the window.

jwilling replied:
At the very least, can you move this out of the whole if-else statement?

```
if(isnan(knobOffset))
    knobOffset = 0.0f;
```

avaidyam replied:
Sure thing.

avaidyam commented:
@jwilling This is good to go now! (Hopefully) I've fixed the poor epileptic scrollers and made sure they animated properly. :sunglasses:

avaidyam commented:
@jwilling And thanks for the video, I seriously did not know what you meant until I saw that. Took me a while to reproduce and fix it, though.

jwilling commented:
Looking much better. There's another problem that should be addressed, though. The resizing of the knobs is animated when you resize the window. The resizing shouldn't be animated. :+1:

avaidyam commented:
hmm, I'll look into that.

avaidyam commented:
@jwilling Uhh, I think I've figured out the general issue, but I can't patch that up without disabling animation completely. Any ideas? It's in `-[TUIScroller layoutSubviews]`...

jwilling commented:
@galaxas0 When is `-layoutSubviews` called? Are you using it as the general updating place for laying out the expanded track as well?

avaidyam commented:
@jwilling I've figured it out and fixed it. :D

jwilling commented:
Okay, we're getting closer. Couple more things I'm noticing now.
- In AppKit, when the window is not key it does not expand the track, even on direct hover.
- Also in AppKit, the sizing of the expanded knob is slightly different.

Notice:

![](http://i.imgur.com/tBIyQ.png)

AppKit uses the exact same padding distance between the right side of the window and the knob, and between the knob and the right side of the darker line going down the track. Then there is an additional bit added on to create the line. In reality, it's actually off center, but it makes more sense that way.

avaidyam committed:
Add knob inset to center scroller and use window keyedness to allow expansion.

avaidyam commented:
@jwilling @jspahrsummers Are there any more discrepancies with AppKit's scrollers?

avaidyam commented:
@jwilling @jspahrsummers :mailbox_with_mail:

jwilling commented:
I'm going to be super picky here. There are a couple more things that bother me a little bit.
- When the expanded track is shown and the mouse isn't touch, it fades away. Great! But when you move the trackpad again and the scrollers are shown, the knob animates from the expanded to the smaller size when it shouldn't.
- When pulling past the bounds of the view, the track should contract in size instead of staying the same. This gives the illusion that you really are pulling past the bounds of the view.
- The duration of the fade seems to be too long. Looking at a sample of apps that appear to be using a default `NSScrollView`, it looks like they have a one-second fade (including the alpha animation to 0) instead of two seconds.

jwilling commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,465 @@
+	if(oldKnobWidth == newKnobWidth)
+		self.knob.frame = frame;
+	
+	[TUIView animateWithDuration:0.25 animations:^{
```
Can this be drawn out into a constant? Should this be `TUIScrollerFadeSpeed`?

avaidyam replied:
Yes, yes it should. :innocent:

jwilling commented on lib/UIKit/TUIScrollView.m:
```diff
@@ -364,140 +385,199 @@ - (CGPoint)_fixProposedContentOffset:(CGPoint)offset
+		self.verticalScroller.frame = newVScrollerRect;
+		self.horizontalScroller.frame = newHScrollerRect;
+	};
+	[TUIView animateWithDuration:0.25 animations:^{
```
Same note as above.

avaidyam replied:
:+1:

avaidyam commented:
@jwilling It's not called being picky, it's called quality control ;) I'm quite happy to fix all this up, as it only makes the overall TwUI experience better if we're picky about the small things.

jwilling commented:
:+1:

avaidyam commented:
@jwilling For your last point, did you mean, exactly 1 second vs 2 seconds? Because that's super long. (I believe it's only 0.25 seconds...)

jwilling commented:
Currently with the fading and all, these scrollers appear to stay on the screen for almost two seconds. Ideally with the fade duration added, the maximum time on screen after scrolling has stopped should be less than or equal to a second.

avaidyam commented:
Oh, the scroller appearance timer. Gotcha! :sparkles:

avaidyam committed:
Change scroller display duration.

jwilling commented:
Aha, I think I know why this discrepancy is happening. When scrolling and letting go, it appears that `TUIScrollView` is correcting non-integral coordinates via a subtle animation, which appears to keep the scrollbars alive longer than they need to. When the scroll view is scrolled to a location with non-fractional coordinates the duration appears to be much better. Are you able to look into the fractional coordinates and see what can be done about that? I'd take a look myself but I'm swamped at the moment. :+1:

avaidyam commented:
@jwilling Alright, I give up-- I have no clue what this effect is (can't see it on my end) nor where it might occur in code. I'll move on to the other two.

jwilling commented:
:+1: It might have something to do with a retina display. I'll see if I can screen capture it when I get the chance.

avaidyam commented:
Thanks! I need me a Retina display MBP soon… sadly, I'm only a high school senior. Can't afford one! :sadface:

jwilling commented:
And I'm "only" a college freshman. :wink:

avaidyam committed:
Don't animate a scroller shrink.

avaidyam committed:
Merge remote-tracking branch 'upstream/master' into upstream-ML-scrollbars

avaidyam committed:
Working overscroll rubber banding!

avaidyam commented:
@jwilling @jspahrsummers I've FINALLY understood the logic behind the rubber-banding! Hurrah! Would you mind checking this out for any other issues?

jwilling commented:
When pulling at the bottom of the view, the shrinking goes the wrong way. It should shrink down instead of shrinking up, which it currently does. The same goes for left / right.

Additionally, the top of the scroller jerks every single frame of the animation, almost like it moves up one pixel before resize, and then down again after resize.

avaidyam committed:
Fixed overscroll on bottom/left edges and scroller jumps. Should look better on Retina (I think!)

avaidyam commented:
@jwilling :mailbox_with_mail: This is identical to AppKit's scrollers now. (As far as I can tell!)

jwilling commented:
Getting there, but not done yet.

Here's TwUI's scrollers:
![](http://i.imgur.com/jsIeo.png)

and AppKit:
![](http://i.imgur.com/wunUL.png)

Notice the perfect overlap on the corners.

Also, AppKit actually shows the scrollers when two fingers are placed on the trackpad. The scrollers remain visible while the fingers are held down.

avaidyam committed:
Fixed scroller position and don't hide knob if scroll view is tracking.

avaidyam commented:
@jwilling Both of these issues should be fixed now!
EDIT: Would it still be in the scope of TwUI to add a category on NSTimer for block support, and use that? That would simplify these two methods a lot...

jwilling commented:
I'm afraid that you misunderstood me. When you hold two fingers down on the trackpad and don't move them at all, AppKit's scrollers stay visible. This doesn't seem to be the case right now in this pull request.

avaidyam commented:
Aaahh, okay, I did indeed misunderstand you. Thanks for the clarification! :)

avaidyam commented:
I believe this can't be amended simply-- `-beginGestureWithEvent:` does not pass the first touch pair to the `TUIScrollView`, and therein lies the problem-- not with the scrollers. Can you suggest a fix?

jwilling commented:
Perhaps it might be impossible. I'm looking at a class-dump of AppKit right now.

avaidyam commented:
Yep, sadly. We could've simulated it with two-tap events, but AppKit doesn't give us those either. /cries

avaidyam commented:
@jwilling Then, is there anything left?

avaidyam commented:
@jwilling :mailbox_with_mail:

jwilling commented:
The scroller overlays are almost there, but not quite. Take a look:

![](http://i.imgur.com/5JUyN.png)

jwilling commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,484 @@
+		[self _updateKnobAlphaWithSpeed:0.0];
+		
+		[[NSNotificationCenter defaultCenter] addObserver:self
+												 selector:@selector(preferredScrollerStyleChanged:)
```
These should be aligned by the colons, as Xcode does by default. Same goes for the observer below.

avaidyam replied:
Xcode did align them by colons...

jwilling replied:
Looks like github.com is giving a different tab length than in Xcode. It's okay then.

avaidyam replied:
Thanks! So aside from the (what I think, is a bad idea) if statement spacing, is this good to go?

jwilling commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,484 @@
+@implementation TUIScroller
+
+- (id)initWithFrame:(CGRect)frame {
+	if((self = [super initWithFrame:frame])) {
```
Space after if.

avaidyam replied:
Wait, was there such a requirement? I, personally, detest spaces after the `if`… I'd find it slightly absurd if an open source project had such requirements (I also prefer same-like bracketing, like `if(condition) {`…)

jwilling replied:
Yes. It is in the conventions.

avaidyam replied:
.... that's.... excuse me as I go open a debate issue for this.

jwilling replied:
I'd have to say I agree with the conventions here. Having no space looks pretty bad.

avaidyam replied:
Just left an issue here, because I think quite conversely: a space looks horrid to me. #programmerdebates https://github.com/github/objective-c-conventions/issues/18

jwilling commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,484 @@
+}
+
+- (void)setHideKnobTimer:(NSTimer *)hideKnobTimer {
+	if(!hideKnobTimer && _hideKnobTimer) {
```
Space after if. The same goes for all the other conditionals in this class.

avaidyam replied:
Another note: quite a lot of TwUI code continues to use `if()`, and @jspahrsummers was perfectly okay with it, if that makes any addition to this...

jwilling replied:
Old code does not reflect what new code should look like.

avaidyam replied:
=twitch twitch= Alright, but I don't think imposing a specific spacing/newline format is a good thing... as I've stated before, I don't like spacing my `if`'s... I don't think I've been told to do so on any other PR either. :\

avaidyam replied:
Alright, well, that debate ended fast. I'll fix the `if ()` statments... :cry:

jwilling commented on lib/UIKit/TUIScroller.m:
```diff
@@ -0,0 +1,484 @@
+		visibility = self.scrollView.horizontalScrollIndicatorVisibility;
+	
+	if(visibility != TUIScrollViewIndicatorVisibleNever) {
+		self.hideKnobTimer = nil;
```
Just setting the timer to nil isn't enough, it should be invalidated as well.

jwilling replied:
Apologies, did not see the custom setter. All's good here. :+1:

avaidyam replied:
Alright! :+1:

avaidyam commented:
@jwilling Alright, aside from the only-ever-so-slightly-bewildering requirement for the if statement bracket spacing, it should look better now.

avaidyam committed:
Style fixes.

avaidyam commented:
@jwilling I've spaced all the if()'s.

dannygreg commented:
Conforming to a style guide is not a bizarre request. It is fairly common on most open (and indeed closed) source projects.

avaidyam commented:
I understand that, but I just can't understand how the spacing on an if statement is supposed to matter. Either way, since the law has been laid down, I'll follow it. :)

jwilling commented on lib/UIKit/TUIScrollView.m:
```diff
@@ -364,140 +385,203 @@ - (CGPoint)_fixProposedContentOffset:(CGPoint)offset
+	[self.horizontalScroller forceDisableExpandedScroller:NO];
+	
+	// Conditionally un-expand one scroller (if both are expanded).
+	if(verticalScrollerWasExpanded && horizontalScrollerIsExpanded)
```
Missed some here.

jwilling commented on lib/UIKit/TUIScrollView.m:
```diff
@@ -364,140 +385,203 @@ - (CGPoint)_fixProposedContentOffset:(CGPoint)offset
 	
-	switch(self.verticalScrollIndicatorVisibility){
+	// Determine the visibility of the vertical scroller.
+	switch(self.verticalScrollIndicatorVisibility) {
```
Also here.

jwilling commented on lib/UIKit/TUIScrollView.m:
```diff
@@ -364,140 +385,203 @@ - (CGPoint)_fixProposedContentOffset:(CGPoint)offset
 	
-	switch(self.horizontalScrollIndicatorVisibility){
+	// Determine the visibility of the horizontal scroller.
+	switch(self.horizontalScrollIndicatorVisibility) {
```
And here.

jwilling commented on lib/UIKit/TUIScrollView.m:
```diff
@@ -364,140 +385,203 @@ - (CGPoint)_fixProposedContentOffset:(CGPoint)offset
+	CGFloat horizontalScrollerStart = 0;
+	CGFloat verticalScrollerOffset = ((hVisible ? verticalScrollerSize : 0) + resizeKnobSize.height);
+	CGFloat horizontalScrollerOffset = ((vVisible ? horizontalScrollerSize : 0) + resizeKnobSize.width);
+	if(![self.class requiresLegacyScrollers]) {
```
Here as well.

avaidyam commented:
@jwilling Done! Took care of whatever was in my PR. :100:

jwilling commented:
Close. Check your switch statements as mentioned above.

jwilling commented on lib/UIKit/TUIScrollView.m:
```diff
@@ -364,140 +385,203 @@ - (CGPoint)_fixProposedContentOffset:(CGPoint)offset
 - (void)setResizeKnobSize:(CGSize)s
 {
-	if(!isAtleastLion)
+	if([self.class requiresLegacyScrollers])
```
Also one here.

avaidyam commented:
@jwilling Oops! Got them.

jwilling commented on lib/UIKit/TUIScrollView.m:
```diff
@@ -1034,7 +1118,7 @@ - (void)endGestureWithEvent:(NSEvent *)event
 		[self _startThrow];
 		
-		if(isAtleastLion) {
+		if([self.class requiresElasticSrolling]) {
```
Here as well.

jwilling commented on lib/UIKit/TUIScrollView.m:
```diff
@@ -1051,7 +1135,7 @@ - (void)scrollWheel:(NSEvent *)event
 		int phase = ScrollPhaseNormal;
 		
-		if(isAtleastLion) {
+		if([self.class requiresElasticSrolling]) {
```
And here.

avaidyam committed:
Style fixes for the rest of the class.

avaidyam commented:
@jwilling Got them... all.... ;___; :mailbox_with_mail:

jwilling commented:
Thanks for your work on this! :+1:

avaidyam commented:
@jwilling Thank you _too much_ for unfettering me! LOL, this was such a painful PR to work on, but it had to be done someday. :smile:

jagreenwood commented:
@galaxas0 This looks awesome. Fantastic job!

avaidyam commented:
@jagreenwood Thanks! Much appreciated! :)
